### PR TITLE
Partial sync

### LINKED ISSUE #1570: Partial Sync: Unify notification system
See https://github.com/realm/product/issues/392 for details

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

Closes #1570, #1568, #1569 

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: lib/browser/results.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ createMethods(Results.prototype, objectTypes.RESULTS, [
     'filtered',
     'sorted',
     'snapshot',
+    'subscribe',
```
Comment (nirinchev) [inline lib/browser/results.js]: Have we reached binding-level agreement on the `subscribe` terminology? My only concern is that it doesn't seem descriptive enough and may be confused with subscribing for notifications.

File: lib/extensions.js [hunk 2]
```diff
@@ -159,18 +159,13 @@ module.exports = function(realmConstructor) {
...
-                        reject(err);
-                    } else {
-                        resolve(results);
-                    }
-                });
-            });
-            return promise;
+        // Keep these value in sync with subscription_state.hpp
+        realmConstructor.Sync.SubscriptionState = {
+            Undefined: -3,         // Unknown which state Partial Sync is in.
```
Comment (nirinchev) [inline lib/extensions.js]: Are these going to be exposed to end-users? If yes, `unknown which state Partial Sync is in` is too vague.

File: lib/extensions.js [hunk 3]
```diff
@@ -159,18 +159,13 @@ module.exports = function(realmConstructor) {
...
+            NotSupported: - 2,     // Partial Sync not supported.
+            Error: -1,             // An error was detect in Partial Sync.
+            Uninitialized: 0,      // The subscription was just created, but not handled by sync yet.
+            Initialized: 1,        // The subscription has been initialized successfully and are syncing data to the dev
```
Comment (nirinchev) [inline lib/extensions.js]: `to the dev`?

File: src/js_collection.hpp [hunk 4]
```diff
@@ -23,6 +23,9 @@
 #include "js_observable.hpp"
 
 #include "collection_notifications.hpp"
+#if REALM_ENABLE_SYNC
```
Comment (nirinchev) [inline src/js_collection.hpp]: Is there a case where the js binding has this set to false?

Comment (kneth) [reply src/js_collection.hpp]: I believe that we test both with and without sync on CI.

File: lib/extensions.js (hunk 3 above)
Comment (kneth) [reply lib/extensions.js]: `ice`

File: lib/extensions.js (hunk 2 above)
Comment (kneth) [reply lib/extensions.js]: Agree, we should document them more carefully them in the API documentation.

File: src/js_collection.hpp (hunk 4 above)
Comment (nirinchev) [reply src/js_collection.hpp]: Seems like wasted cycles if we don't release it after testing 😛

File: lib/browser/results.js (hunk 1 above)
Comment (bdash) [reply lib/browser/results.js]: I don't think there's been any serious discussion about the specific naming used here. If you're concerned about this naming, we have very little time to resolve it, so I'd suggest bringing it up in Slack so we can hammer it out promptly.

File: src/js_collection.hpp (hunk 4 above)
Comment (bdash) [reply src/js_collection.hpp]: Isn't realm-js on Linux built without sync unless you're installing one of the paid editions?

Comment (nirinchev) [reply src/js_collection.hpp]: I don't think so - we have a single package and feature tokens to unlock functionality.

File: src/js_realm.hpp [hunk 5]
```diff
@@ -248,7 +248,7 @@ class RealmClass : public ClassDefinition<T, SharedRealm, ObservableClass<T>> {
         {"_objectForObjectId", wrap<object_for_object_id>},
  #if REALM_ENABLE_SYNC
         {"_waitForDownload", wrap<wait_for_download_completion>},
-        {"_subscribeToObjects", wrap<subscribe_to_objects>},
+        {"getQueryStatus", wrap<get_query_status>},
```
Comment (bdash) [inline src/js_realm.hpp]: Is this used for anything? I cannot see any references to `get_query_status`.

File: src/js_results.hpp [hunk 6]
```diff
@@ -254,6 +263,31 @@ void ResultsClass<T>::is_valid(ContextType ctx, ObjectType this_object, Argument
...
+#if REALM_ENABLE_SYNC
+template<typename T>
+void ResultsClass<T>::subscribe(ContextType ctx, ObjectType this_object, Arguments args, ReturnValue &return_value) {
+    args.validate_maximum(1);
+
+    auto results = get_internal<T, ResultsClass<T>>(this_object);
+    auto realm = results->get_realm();
+    auto sync_config = realm->config().sync_config;
+    if (!sync_config || !sync_config->is_partial) {
+        throw std::logic_error("A partial sync query can only be registered in a partially synced Realm.");
```
Comment (bdash) [inline src/js_results.hpp]: `partial_sync::subscribe` performs this check internally, so I don't think it's necessary to repeat it here?

File: docs/sync.js [hunk 7]
```diff
@@ -468,6 +468,58 @@ class Session {
...
+
+    /**
+     * Gets the error message. `undefined` if no error.
+     * @type {string}
+     */
+    get error() {}
+
+    /**
+     * Unsubscribe a partial synced `Realm.Results`. The state will change to `Realm.Sync.SubscriptionState.Invalidated`.
+     * The `Realm.Results` will not produce any meaningful values.
```
Comment (bdash) [inline docs/sync.js]: It'd be worth mentioning here that any objects matching this query will be removed if they are not matched by any other query, and that they'll be removed asynchronously with respect to this method.

File: docs/sync.js [hunk 8]
```diff
@@ -468,6 +468,58 @@ class Session {
...
+    unsubscribe() {}
+
+    /**
+     * Adds a listener `callback` which will be called when the state of the subscription changes.
+     * @param {function(state)} callback - A function to be called when changes occur.
+     * @throws {Error} If `callback` is not a function.
+     * @example
+     * subscription.addListener((subscription, state) => {
+     *     if (state == Realm.Sync.SubscriptionState.Complete) {
+     *         var partial_results = subscription.results;
```
Comment (bdash) [inline docs/sync.js]: I think you removed the `results` property for now?

**Pull request merged**